cpp/arduino/SPI.h

```diff
@@ -94,10 +98,7 @@ class SPIClass: public ObservableDataStream {
   uint16_t transfer16(uint16_t data) {
     union { uint16_t val; struct { uint8_t lsb; uint8_t msb; }; } in, out;
     in.val = data;
-    // Changes in the definition of _SFR_IO8() cause this to break.
-    // This horible hack allows the tests to pass but is done 
-    // without any understanding at all of what is going on here!
-    if (false && !(SPCR & (1 << DORD))) {
+    if (bitOrder == MSBFIRST) {
```

Comment:
I don't think this is correct. AFAIU, `MSBFIRST` / `LSBFIRST` are about the  bit order, so the order of bits within a byte. This if is about byte order, so the order of bytes within a 16-bit word. AFAIK, the official Arduino API lacks a way to configure this byte order, but on AVR the hardware handles this, so sketches and libraries can mess with the `SPCR` register to let the hardware reverse the bit order.

Reply:
You are right, the flags are for bits not bytes. I'll follow your earlier suggestion of setting and checking the `SPCR` register and try to build some tests. Also, that will require following your (even) earlier code of checking for `SPCR` and `DORD` being defined. It will be a bit more complicated, but I'll give it a go!

Reply:
Further investigation suggests that `DORD` is the _bit order_, not _byte order_ (see [here](http://avrbeginners.net/architecture/spi/spi.html) and [here](https://ww1.microchip.com/downloads/en/AppNotes/Atmel-2585-Setup-and-Use-of-the-SPI_ApplicationNote_AVR151.pdf)). So, I'm again confused. 

Maybe if you are doing most significant bit first then you need to get the most significant byte first and if you are doing the least significant bit first then you need to start with the least significant byte. In that case the code is okay? But it could be improved by using `SPCR` and `DORD`.

Reply:
Right, maybe I was working too much off assumptions based on the mocked SPI.h :-)

So then I guess `DORD` actually decides the bit order, which also implies a particular the byte order. And since the mocked SPI interface is not a physical one where bit order plays a role, the bit order is actually not so relevant (since we just store stransmitted bytes in memory), but the byte order is (since we split a 16-bit word into two bytes).

Looking at the actual SPI implementation (https://github.com/arduino/ArduinoCore-avr/blob/master/libraries/SPI/src/SPI.h) it seems that, unlike what I assumed, the hardware does not actually implement 16-bit transfers, this is done in software. So the official AVR `transfer16` also checks `DORD` to see whether it should send the MSByte or LSByte first (so basically it just lets the byte order follow the configured bit order).

Then, what I said before about there not being an API to set the byte order, so sketches would write `DORD` also seems to be untrue: The byte order always follows the bit order, so the byte order can be set by setting the bit order. Even more, the AVR SPI code checks `DORD` in `transfer16`, but `DORD` is set in e.g. `beginTransaction` based on `MSBFIRST` / `LSBFIRST`. IOW, they just check `DORD` in `transfer16` so they don't need to store `MSBFIRST` / `LSBFIRST` anywhere.


Summarizing: I was quite wrong about nearly everything, and your [previous fix that remembered `bitOrder`](https://github.com/Arduino-CI/arduino_ci/pull/201/commits/57070abdd6cdeca7ac68c391dccd68aa5963b1c5) actually seems like the correct one (so I would suggest reverting the SPI-related changes in your most recent commit).

An unrelated note: Your last commit contains two fairly separate changes, it would be easier to review if they were in separate commits (in git, the first line is considered the title and is the only one shown in a lot of places, so I initially thought that you maybe did not push your SPI changes yet). Also, if these commits are merged as-is, having such a double commit in the official history makes it harder to look through the commit history later (I'd personally restructure them, squashing multiple revisions of the same fix and separating out different fixes before merging, though).

Reply:
I've reverted the `bitOrder` code and will keep in mind the commit history suggestions. Thanks for the feedback!